### ISSUE TITLE
Building without depending on X11 on Raspberry Pi

### ISSUE THREAD

**Issue by vanfanel** [lunixbochs/glshim#158]
Hi,

First of all, thanks a lot for your work on this! I have yet to see ghshim working but I have heard awesome things about it, specially for hardware without OpenGL 1.x implementations like the Pi.

My understand is that glshim DOES work on the Pi with the LIBGL_FB=1 runtime parameter, so it doesn't really need X at all to run. Also, seein the dispmanx parts of the code confirms my idea :)
And there is this, too: https://github.com/lunixbochs/glshim/issues/23
So, all in all, X is not a real dependency.
I haven't the huge X dependencies on my minimal Raspbian system, and I am trying to build glshim, but I get it looking for the X libs anyway at build time, so I get:

 fatal error: X11/Xlib.h: No such file or directory

Well, since X11 is not a dependency for runtime, could it be removed as a dependency at build time, too?
After all, it's already working in that situation, isn't it?

**Comment by lunixbochs** [comment]
you need X libs to build glshim until #34 is done

it doesn't actually need them if you don't have an X server, and you should be able to hand-remove any traces of X dependencies without breaking glshim for your own build

**Comment by MagaTailor** [comment]
@vanfanel Actually, it got hardware OpenGL 2.0 support about a month ago - not sure if the original Rpi is also supported by the new driver, but it probably should as the GPU is the same.

**Comment by vanfanel** [comment]
@lunixbochs @petevine , thanks for your responses!
So, you say I can remove the X headers and X-relared struct fields etc, without breaking glshim? Why don't you simply ifdef these parts in the code then?

Edit: I have tried commenting out the X related headers (Xlib and Xutils) and it seems to be needing the Display data type in a LOT of places, so it won't build...

**Comment by vanfanel** [comment]
@lunixbochs: can you give me any indications on this, please? (if you have time)

**Comment by vanfanel** [comment]
@lunixbochs I have spent hours trying to remove the X11 dependencies as per these instructions and I got really annoyed. No way it works for me: there are thousands of errors after removing the X11 headers and functions, and typedef-ing Display and the rest to void (I did so in include/GL/glx.h).
I am NOT asking you to do it for me: I have tried several times and failed, but since you know the code well you could simply put some ifdefs around the parts that are X11 dependant so we can finally build without X11.

**Comment by lunixbochs** [comment]
So in my gcw0 branch I actually kept the X11 headers and just removed all X11 function calls and linkage - https://github.com/lunixbochs/glshim/commit/a6c8f831615a190c1fd04a9147447a66703b8abf

I haven't played with it since. It's not as simple as putting ifdefs around the X11-dependent code, because that's the entire GLX API. Games won't load without GLX, as it's how they set up the OpenGL context.

**Comment by vanfanel** [comment]
@lunixbochs There are no X11 headers in minimal Raspbian systems :(
That's precissely the problem. No X11 headers, no X11 functions, no X11 types.
After all this is an OpenGL -> OpenGL_ES translator, why keep X11 dependencies?  In most cases, there will be no X11 headers or libs.

**Comment by lunixbochs** [comment]
The ABI _requires_ X11 compatible functions. GLX is an X11 ABI. It's not just the glshim headers - the GLX headers include X11 as well. You don't need the libraries. Just the headers. You have apt-get. You can install libx11-dev without installing X11 itself.

I'll work around this in the long term.

**Comment by vanfanel** [comment]
@lunixbochs So the X11 headers will always be needed? I don't quite understand... In theory, glshim uses EGL for context management, GLES for rendering and that's it.

**Comment by lunixbochs** [comment]
The problem isn't the backend, it's the ABI-compatible GLX/X11 frontend glshim exposes to applications.  You don't need to recompile games to use them with glshim. Most of these games were originally built against the same X11 and GLX headers you're complaining that glshim uses. If you have a game that doesn't use the GLX backend, you can go ahead and remove the entire GLX folder from glshim and it should continue to work. This, however, will be effectively zero of the games using OpenGL 1.x.

I use the X11 and GLX headers _right now_ to make that interface work. I just said I'd work around this in the long term, and there's an open issue to not require X11 at compile time.

If you're not comfortable making these source modifications, all you should need to do is install the X11 client libraries via `sudo apt-get install libx11-dev`.

When installing libx11-dev on my local machine: `After this operation, 20.7 MB of additional disk space will be used.`

Until further notice, glshim requires the _client libraries and headers only_ of X11 to build, due to the GLX interface it exposes to games that were also built against X11. The X11 client development package is very small (~20MB) and should be more than reasonable to install when compared to most games. glshim does not require an active (or even installed) X11 server to run, though some games might.

Closing as a duplicate of #34.
